crates/engine/tree/src/tree/mod.rs

```diff
@@ -806,15 +817,46 @@ where
     ///
     /// This mainly compares the missing parent of the downloaded block with the current canonical
     /// tip, and decides whether or not backfill sync should be triggered.
-    const fn on_disconnected_downloaded_block(
+    fn on_disconnected_downloaded_block(
         &self,
         downloaded_block: BlockNumHash,
         missing_parent: BlockNumHash,
         head: BlockNumHash,
     ) -> Option<TreeEvent> {
-        None
+        // compare the missing parent with the canonical tip
+        if let Some(target) =
+            self.backfill_sync_target(head.number, missing_parent.number, Some(downloaded_block))
+        {
+            return Some(TreeEvent::BackfillAction(BackfillAction::Start(target.into())));
+        }
+
+        // continue downloading the missing parent
+        //
+        // this happens if either:
+        //  * the missing parent block num < canonical tip num
+        //    * this case represents a missing block on a fork that is shorter than the canonical
+        //      chain
+        //  * the missing parent block num >= canonical tip num, but the number of missing blocks is
+        //    less than the pipeline threshold
+        //    * this case represents a potentially long range of blocks to download and execute
+        let request = if let Some(distance) =
+            self.distance_from_local_tip(head.number, missing_parent.number)
+        {
+            DownloadRequest::BlockRange(missing_parent.hash, distance)
+        } else {
+            // This happens when the missing parent is on an outdated
+            // sidechain and we can only download the missing block itself
+            DownloadRequest::BlockSet(HashSet::from_iter([missing_parent.hash]))
+        };
```

Comment:
Makes sense, I like documenting these outcomes explicitly, since the `Option` alone doesn't make the meaning super clear.